File: app-java/src/main/java/tech/thdev/java_udemy_sample/adapter/ImageSampleAdapter.java
package tech.thdev.java_udemy_sample.adapter;

import android.content.Context;
import android.view.ViewGroup;

import androidx.recyclerview.widget.RecyclerView;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;

import tech.thdev.java_udemy_sample.adapter.holder.ImageSampleViewHolder;
import tech.thdev.java_udemy_sample.data.ImageItem;

/**
 * Created by tae-hwan on 10/29/16.
 */

public class ImageSampleAdapter extends RecyclerView.Adapter<ImageSampleViewHolder> {

    private final Context context;
    private ArrayList<ImageItem> imageItems = new ArrayList<>();

    public ImageSampleAdapter(Context context) {
        this.context = context;
    }

    @NotNull
    @Override
    public ImageSampleViewHolder onCreateViewHolder(@NotNull ViewGroup parent, int viewType) {
        return new ImageSampleViewHolder(context, parent);
    }

    @Override
    public void onBindViewHolder(ImageSampleViewHolder holder, int position) {
        holder.onBindView(imageItems.get(position), position);
    }

    @Override
    public int getItemCount() {
        return imageItems.size();
    }

    public void addItem(ImageItem imageItem) {
        imageItems.add(imageItem);
    }
}

File: app-kotlin/src/main/java/tech/thdev/kotlin_udemy_sample/adapter/ImageSampleAdapter.kt
package tech.thdev.kotlin_udemy_sample.adapter

import android.content.Context
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import tech.thdev.kotlin_udemy_sample.adapter.holder.ImageSampleViewHolder
import tech.thdev.kotlin_udemy_sample.data.ImageItem

/**
 * Created by tae-hwan on 10/29/16.
 */

class ImageSampleAdapter(private val context: Context) : RecyclerView.Adapter<ImageSampleViewHolder>() {

    val itemList: MutableList<ImageItem> = ArrayList()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ImageSampleViewHolder {
        return ImageSampleViewHolder(context, parent)
    }

    override fun onBindViewHolder(holder: ImageSampleViewHolder, position: Int) {
        holder?.bindView(getItem(position), position)
    }

    /**
     * Item list의 item을 return
     */
    private fun getItem(position: Int) =
        itemList[position]

    /**
     * Item size를 return
     */
    override fun getItemCount() =
        itemList.size

    fun addItem(it: ImageItem) {
        itemList.add(it)
    }
}